File: frontend/src/types/subscription.ts
import type { CategoryResponseDTO } from "./category";
import type { TransactionType } from "./transaction";

export type SubscriptionStatus = 'ACTIVE' | 'PAUSED' | 'CANCELLED';
export type BillingPeriod = 'WEEKLY' | 'MONTHLY' | 'YEARLY';

export interface SubscriptionRequestDTO {
  title: string;
  amount: number;
  currency: string;
  categoryId: number;
  billingPeriod: BillingPeriod;
  nextBillingPeriod: string;
  status: SubscriptionStatus;
  transactionType: TransactionType;
}
export interface SubscriptionResponseDTO {
  id: number;
  title: string;
  amount: number;
  currency: string;
  category: CategoryResponseDTO;
  billingPeriod: BillingPeriod;
  nextBillingPeriod: string;
  status: SubscriptionStatus;
  transactionType: TransactionType;
}

File: backend/src/main/java/com/home/finance_tracker/subscription/entity/Subscription.java
package com.home.finance_tracker.subscription.entity;

import com.home.finance_tracker.category.entity.Category;
import com.home.finance_tracker.transaction.entity.TransactionType;
import com.home.finance_tracker.user.entity.User;
import jakarta.persistence.*;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.math.BigDecimal;
import java.time.LocalDate;

@Entity
@Table(name = "subscriptions")
@Getter
@Setter
@NoArgsConstructor
public class Subscription {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    private Long id;

    @Column(nullable = false, length = 150)
    private String title;

    @Column(nullable = false, precision = 19, scale = 2)
    private BigDecimal amount;

    @Column(nullable = false, length = 3)
    private String currency = "PLN";

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "category_id", nullable = false)
    private Category category;

    @Enumerated(EnumType.STRING)
    @Column(name = "billing_period", length = 20)
    private BillingPeriod billingPeriod;

    @Column(name = "next_billing_period", nullable = false)
    private LocalDate nextBillingPeriod;

    @Enumerated(EnumType.STRING)
    @Column(length = 20, nullable = false)
    private SubscriptionStatus status = SubscriptionStatus.ACTIVE;

    @Enumerated(EnumType.STRING)
    @Column(name = "transaction_type", nullable = false, length = 20)
    private TransactionType transactionType;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = true)
    private User user;
}
